### PR TITLE
[SPARK-48788][CORE][UI] Expose task peak onheap/offheap execution memory to API and Spark UI

### PR DESCRIPTION
**Pull request by liuzqt**
### What changes were proposed in this pull request?

https://github.com/apache/spark/pull/47776 has introduced task peak on/off heap execution memory metrics, this PR exposes these two metrics to APIs, also shown in Spark UI Stage page, specifically, in 3 relevant sections(see screen shot below):
- stage details
- task metrics summary (quantiles)
- task metrics

### Does this PR introduce _any_ user-facing change?

Yes. Expose metrics to APIs, also show in Spark UI Stage page.

### How was this patch tested?

Existing UTs.

Manually verify through Spark UI:
**- html page**
[Spark shell - Details for Stage 4 (Attempt 0).mhtml.zip](https://github.com/user-attachments/files/16791980/Spark.shell.-.Details.for.Stage.4.Attempt.0.mhtml.zip)
**- Stage details**
Screenshot 2024-08-29 at 10 45 20 AM
**- Task metrics summary**
Screenshot 2024-08-29 at 10 44 37 AM
**- Task metrics**
Screenshot 2024-08-29 at 10 45 06 AM

### Was this patch authored or co-authored using generative AI tooling?
NO

### CODE REVIEW

**Comment by liuzqt** [comment]
Hi @mridulm @dongjoon-hyun I'm working on this follow-up item for https://github.com/apache/spark/pull/47776, mostly APIs change and Spark UI change, I've manually verified it in Spark UI, could you pls help review this when you have time? 

And BTW do you know how to run `HistoryServerSuite.main`? seems like I need to run that to re-generate the ground truth.

File: core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto [hunk 1]
```diff
@@ -128,6 +128,8 @@ message TaskDataWrapper {
   int64 shuffle_merged_local_bytes_read = 49;
   int64 shuffle_remote_reqs_duration = 50;
   int64 shuffle_merged_remote_req_duration = 51;
+  int64 peak_on_heap_execution_memory = 52;
+  int64 peak_off_heap_execution_memory = 53;
```
Comment (mridulm) [inline core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto:132]: Add to `ExecutorStageSummary`, `ExecutorSummary`, `ExecutorMetricsDistributions` as well.
(Here and other model classes)

**Review by mridulm — COMMENTED** [review]
Took a quick look, but I am not very familiar with this side of spark.

+CC @yaooqinn, @yanboliang as well for review.

**Comment by mridulm** [comment]
Also +CC @gengliangwang

File: core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto (hunk 1 above)
Comment (liuzqt) [reply core/src/main/protobuf/org/apache/spark/status/protobuf/store_types.proto:132]: Looks like we already have executor level stage memory metrics:
- `ExecutorStageSummary` has `val peakMemoryMetrics: Option[ExecutorMetrics]`
- `ExecutorSummary` has `val memoryMetrics: Option[MemoryMetrics]`
- `ExecutorMetricsDistributions` has `val peakMemoryMetrics: ExecutorPeakMetricsDistributions`

which are aggregated through [AppStatusListener.updateStageLevelPeakExecutorMetrics](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/status/AppStatusListener.scala#L1000)

**Comment by liuzqt** [comment]
Hi @mridulm I've fixed broken tests, also answered you questions, do you mind taking another look when you have time? Thanks

**Pull request closed without merging**